// Gus/Command/CommandContext.ts
import * as Discord from 'discord.js'

export default class CommandContext
{
    server: Discord.Guild
    message: Discord.Message
    channel: Discord.TextChannel | Discord.DMChannel | Discord.GroupDMChannel
    reactions: Discord.Collection<string, Discord.MessageReaction>
    executor: Discord.User

    /**
     * Constructor of the CommandContext class.
     *
     * @classdesc The command context
     */
    constructor(message: Discord.Message)
    {
        this.server = message.guild
        this.channel = message.channel
        this.message = message
        this.reactions = message.reactions
        this.executor = message.author
    }

    reply(content?: any, options?: Discord.MessageOptions | Discord.RichEmbed | Discord.Attachment)
    {
        if (typeof this.channel !== 'undefined')
            return this.channel.send(content, options)
        else
            return this.executor.send(content, options)
    }
    
    replyEmbed(message, embed) {
        // TODO: Permission error handler
        return this.reply(message, embed)        
    }

    replyError(message, embed) {
        // TODO: Pre-defined errors
        // TODO: Error generator
        return this.replyEmbed(message, embed)
    }
}
