// Gus/DB/Global.ts
import * as low from 'lowdb'
import * as FileSync from 'lowdb/adapters/FileSync'
import { GuildMember } from 'discord.js'

export default class GlobalDB
{
    private dbPath: string = './'
    private adapter: FileSync
    private db: low

    /**
     * Hydrate the class with the database
     * @param dbPath Path of database json file
     */
    constructor(dbPath)
    {
        this.dbPath += dbPath

        this.adapter = new FileSync(this.dbPath, {defaultValue: {}})
        this.db = low(this.adapter)

        this.db.defaults(
            {
                servers: [],
                globalConfig: {}
            }).write()
    }

    /**
     * Get all users of a server
     * @param serverid ID of the server
     */
    getUsers(serverid: string) {
        return this.db.get('servers').find({id: serverid}).get('users')
    }

    /**
     * Get advanced config of a server
     * @param serverid ID of the server
     */
    getConfig(serverid: string) {
        return this.db.get('servers').find({id: serverid}).get('config')
    }

    /**
     * Get data of a specific user
     * @param serverid ID of the server
     * @param userid ID of the user
     */
    getUser(serverid: string, userid: string) {
        return this.getUsers(serverid).find({id: userid})
    }
}
