// Gus/DB/User.ts
/*
Rank Equations :
    X = 25L² - 25L
    L = Math.floor(25 + Math.sqrt(625 + 100 * xp)) / 50
*/

import { GuildMember } from 'discord.js'

export default class UserDB
{
    private user : GuildMember
    private userData : any
    xp : number 

    /**
     * Hydrate the class with user data 
     * @param {Discord.GuildMember} user The user to  
     */
    constructor(user : GuildMember)
    {
        this.user = user
        this.userData = app.db.getUser(user.guild.id, user.id)
        this.xp = this.userData.get('xp', 0).value()
    }

    /**
     * Give xp to the user
     * @param {number} quantity The quantity of xp to give to the user 
     */
    incrementXp(quantity: number) {
        this.xp += quantity
        return this.xp
    }

    /**
     * Get level from XP
     * @param xp Quantity of XP
     */
    getLevel(xp : number = this.xp) {
        return Math.floor(Math.floor(25 + Math.sqrt(625 + 100 * xp)) / 50)
    }

    /**
     * Get current level XP
     */
    getXP() {
        return this.xp - this.getLevelRequiredXP()
    }

    /**
     * Get total XP required to reach level
     * @param level Level
     */
    getLevelRequiredXP(level : number = this.getLevel()) {
        return 25 * level * level - 25 * level
    }

    /**
     * Update user's data in database
     */
    push() {
        this.userData.set('xp', this.xp).write()
    }
}
