// Gus/DB/User.ts
import { GuildMember } from 'discord.js'

export default class UserDB
{
    private user : GuildMember
    private userData : any
    xp : number
    level : number

    /**
     * Hydrate the class with user data 
     * @param {Discord.GuildMember} user The user to  
     */
    constructor(user : GuildMember)
    {
        this.user = user
        this.userData = app.db.getUser(user.guild.id, user.id)
        this.xp = this.userData.get('xp', 0).value()
        this.level = this.userData.get('level', 0).value()
    }

    /**
     * Give xp to the user
     * @param {number} quantity The quantity of xp to give to the user 
     */
    incrementXp(quantity: number) {
        this.xp += quantity
    }

    /**
     * Augment user's level
     * @param quantity Number of level 
     */
    incrementLevel(quantity : number = 1) {
        this.level += quantity
    }

    /**
     * Update user's data in database
     */
    push() {
        this.userData.set('xp', this.xp).write()
        this.userData.set('level', this.level).write()
    }
}
